File: client/client-multi/client-java-core/src/test/java/org/kaaproject/kaa/client/event/registration/DefaultEndpointRegistrationManagerTest.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.event.registration;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyListOf;
import static org.mockito.Matchers.anyMapOf;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;
import org.kaaproject.kaa.client.channel.UserTransport;
import org.kaaproject.kaa.client.channel.impl.transports.DefaultUserTransport;
import org.kaaproject.kaa.client.context.ExecutorContext;
import org.kaaproject.kaa.client.event.EndpointAccessToken;
import org.kaaproject.kaa.client.event.EndpointKeyHash;
import org.kaaproject.kaa.client.persistence.KaaClientState;
import org.kaaproject.kaa.common.endpoint.gen.EndpointAttachResponse;
import org.kaaproject.kaa.common.endpoint.gen.EndpointDetachResponse;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponseResultType;
import org.kaaproject.kaa.common.endpoint.gen.UserAttachNotification;
import org.kaaproject.kaa.common.endpoint.gen.UserAttachResponse;
import org.kaaproject.kaa.common.endpoint.gen.UserDetachNotification;
import org.kaaproject.kaa.common.endpoint.gen.UserSyncResponse;
import org.mockito.Mockito;
import org.springframework.test.util.ReflectionTestUtils;

public class DefaultEndpointRegistrationManagerTest {

    private static final int REQUEST_ID = 42;
    
    private static ExecutorContext executorContext;
    private static ExecutorService executor;
    
    @BeforeClass
    public static void beforeSuite(){
        executorContext = Mockito.mock(ExecutorContext.class);
        executor = Executors.newSingleThreadExecutor();
        Mockito.when(executorContext.getApiExecutor()).thenReturn(executor);
        Mockito.when(executorContext.getCallbackExecutor()).thenReturn(executor);
    }
    
    @AfterClass
    public static void afterSuite(){
        executor.shutdown();
    }
    
    @Test(expected=IllegalStateException.class)
    public void checkUserAttachWithNoDefaultVerifier() throws Exception {
        KaaClientState state = mock(KaaClientState.class);

        when(state.getEndpointAccessToken()).thenReturn("");

        UserTransport transport = mock(UserTransport.class);
        DefaultEndpointRegistrationManager manager = spy(new DefaultEndpointRegistrationManager(state, executorContext, transport, null));

        manager.attachUser("externalId", "token", null);
    }
    
    @Test
    public void checkUserAttachWithCustomVerifier() throws Exception {
        KaaClientState state = mock(KaaClientState.class);
        ExecutorContext executorContext = Mockito.mock(ExecutorContext.class);
        when(state.getEndpointAccessToken()).thenReturn("");

        UserTransport transport = mock(UserTransport.class);
        DefaultEndpointRegistrationManager manager = spy(new DefaultEndpointRegistrationManager(state, executorContext, transport, null));

        manager.attachUser("verifierToken", "externalId", "token", null);
        verify(transport, times(1)).sync();
    }

    @Test
    public void checkAttachEndpoint() throws Exception {
        KaaClientState state = mock(KaaClientState.class);
        ExecutorContext executorContext = Mockito.mock(ExecutorContext.class);
        when(state.getEndpointAccessToken()).thenReturn("");
        UserTransport transport = mock(UserTransport.class);
        DefaultEndpointRegistrationManager manager = spy(new DefaultEndpointRegistrationManager(state, executorContext, transport, null));
        OnAttachEndpointOperationCallback listener = mock(OnAttachEndpointOperationCallback.class);
        manager.attachEndpoint(new EndpointAccessToken("accessToken1"), listener);
        manager.attachEndpoint(new EndpointAccessToken("accessToken2"), listener);

        ReflectionTestUtils.setField(manager, "userTransport", null);
        manager.attachEndpoint(new EndpointAccessToken("accessToken3"), null);
        verify(transport, times(2)).sync();
    }

    @Test
    public void checkDetachEndpoint() throws Exception {
        KaaClientState state = mock(KaaClientState.class);
        ExecutorContext executorContext = Mockito.mock(ExecutorContext.class);
        when(state.getEndpointAccessToken()).thenReturn("");

        OnDetachEndpointOperationCallback listener = mock(OnDetachEndpointOperationCallback.class);
        UserTransport transport = mock(UserTransport.class);

        DefaultEndpointRegistrationManager manager = spy(new DefaultEndpointRegistrationManager(state, executorContext, transport, null));
        manager.detachEndpoint(new EndpointKeyHash("keyHash1"), listener);
        manager.detachEndpoint(new EndpointKeyHash("keyHash2"), listener);

        ReflectionTestUtils.setField(manager, "userTransport", null);
        manager.detachEndpoint(new EndpointKeyHash("keyHash3"), null);
        verify(transport, times(2)).sync();
    }

    @Ignore("Ignore during removing endpoint list change listener .")
    @Test
    public void checkEndpointAttachDetachResponse() throws Exception {
        try {
            KaaClientState state = mock(KaaClientState.class);
            ExecutorContext executorContext = Mockito.mock(ExecutorContext.class);
            when(state.getEndpointAccessToken()).thenReturn("");

            UserTransport transport = new DefaultUserTransport();
            DefaultEndpointRegistrationManager manager = spy(new DefaultEndpointRegistrationManager(state, executorContext, transport, null));

            ChangedAttachedEndpointListCallback listListener = mock(ChangedAttachedEndpointListCallback.class);
//            manager.addAttachedEndpointListChangeListener(listListener);

            transport.setEndpointRegistrationProcessor(manager);

            UserSyncResponse sr = new UserSyncResponse();
            sr.setEndpointAttachResponses(null);
            sr.setEndpointDetachResponses(null);

            List<EndpointAttachResponse> attach = new LinkedList<EndpointAttachResponse>();
            attach.add(new EndpointAttachResponse(REQUEST_ID, "keyHash", SyncResponseResultType.SUCCESS));
            sr.setEndpointAttachResponses(attach);
            transport.onUserResponse(sr);

//            manager.removeAttachedEndpointListChangeListener(listListener);

            List<EndpointDetachResponse> detach = new LinkedList<EndpointDetachResponse>();
            detach.add(new EndpointDetachResponse(REQUEST_ID, SyncResponseResultType.SUCCESS));
            sr.setEndpointAttachResponses(null);
            sr.setEndpointDetachResponses(detach);
            transport.onUserResponse(sr);

            verify(manager, times(2)).onUpdate(anyListOf(EndpointAttachResponse.class), anyListOf(EndpointDetachResponse.class), any(UserAttachResponse.class), any(UserAttachNotification.class), any(UserDetachNotification.class));
            verify(listListener, times(1)).onAttachedEndpointListChanged(anyMapOf(EndpointAccessToken.class, EndpointKeyHash.class));
        } catch (IOException e) {
            assertTrue("Unexpected exception " + e.getMessage(), false);
        }
    }

    @Ignore("Ignore during removing endpoint list change listener .")
    @Test
    public void checkAccessTokenChange() {
        KaaClientState state = mock(KaaClientState.class);
        ExecutorContext executorContext = Mockito.mock(ExecutorContext.class);
        
        when(state.getEndpointAccessToken()).thenReturn("token1");

        DefaultEndpointRegistrationManager manager = spy(new DefaultEndpointRegistrationManager(state, executorContext, null, null));

        String accessToken1 = state.getEndpointAccessToken();

        state.refreshEndpointAccessToken();

        String accessToken2 = state.getEndpointAccessToken();

        assertNotEquals("Endpoint access token is same after regeneration!", accessToken1, accessToken2);
    }

    @Test
    public void checkAttachUser() {
        ExecutorContext executorContext = Mockito.mock(ExecutorContext.class);
        KaaClientState state = mock(KaaClientState.class);
        when(state.getEndpointAccessToken()).thenReturn("");

        UserTransport transport = mock(UserTransport.class);
        EndpointRegistrationManager manager = new DefaultEndpointRegistrationManager(state, executorContext, transport, null);
        manager.attachUser("externalId", "userExternalId", "userAccessToken", new UserAttachCallback() {
            @Override
            public void onAttachResult(UserAttachResponse response) {
            }
        });
        verify(transport, times(1)).sync();
    }

    @Test
    public void checkWrappers() {
        String token1 = "token1";
        String token2 = "token2";

        EndpointAccessToken at1 = new EndpointAccessToken(token1);
        EndpointAccessToken at1_2 = new EndpointAccessToken(token1);
        EndpointAccessToken at2 = new EndpointAccessToken(token2);

        assertEquals("EnndpointAccessToken != EndpointAccessToken", at1, at1);
        assertNotEquals("EndpointAccessToken should be not equal to String object", at1, token1);
        assertEquals("toString() returned different value from getToken()", at1.getToken(), at1.toString());
        assertEquals("Objects with equal tokens are not equal", at1, at1_2);
        assertNotEquals("Objects with different tokens are equal", at1, at2);
        assertEquals("Objects' hash codes with equal tokens are not equal", at1.hashCode(), at1_2.hashCode());
        assertNotEquals("Objects' hash codes with different tokens are equal", at1.hashCode(), at2.hashCode());

        at1_2.setToken(token2);
        assertEquals("Objects with equal tokens are not equal", at1_2, at2);
        assertNotEquals("Objects with different tokens are equal", at1, at1_2);

        EndpointAccessToken emptyToken1 = new EndpointAccessToken(null);
        EndpointAccessToken emptyToken2 = new EndpointAccessToken(null);

        assertEquals("Empty objects with are not equal", emptyToken1, emptyToken2);
        assertEquals("Objects' hash codes with empty tokens are not equal", emptyToken1.hashCode(), emptyToken1.hashCode());
        assertNotEquals("Different objects are equal", at1, emptyToken1);
        assertNotEquals("Null-equality of EndpointAccessToken", at1, null);

        String hash1 = "hash1";
        String hash2 = "hash2";

        EndpointKeyHash eoh1 = new EndpointKeyHash(hash1);
        EndpointKeyHash eoh1_2 = new EndpointKeyHash(hash1);
        EndpointKeyHash eoh2 = new EndpointKeyHash(hash2);

        assertEquals("EndpointKeyHash != EndpointKeyHash", eoh1, eoh1);
        assertNotEquals("EndpointKeyHash should be not equal to String object", eoh1, hash1);
        assertEquals("toString() returned different value from getKeyHash()", eoh1.getKeyHash(), eoh1.toString());
        assertEquals("Objects with equal keyHashes are not equal", eoh1, eoh1_2);
        assertNotEquals("Objects with different keyHashes are equal", eoh1, eoh2);
        assertEquals("Objects' hash codes with equal keyHashes are not equal", eoh1.hashCode(), eoh1_2.hashCode());
        assertNotEquals("Objects' hash codes with different keyHashes are equal", eoh1.hashCode(), eoh2.hashCode());

        eoh1_2.setKeyHash(hash2);
        assertEquals("Objects with equal keyHashes are not equal", eoh1_2, eoh2);
        assertNotEquals("Objects with different keyHashes are equal", eoh1, eoh1_2);

        EndpointKeyHash emptyHash1 = new EndpointKeyHash(null);
        EndpointKeyHash emptyHash2 = new EndpointKeyHash(null);

        assertEquals("Empty objects with are not equal", emptyHash1, emptyHash2);
        assertEquals("Objects' hash codes with empty hashes are not equal", emptyHash1.hashCode(), emptyHash1.hashCode());
        assertNotEquals("Different objects are equal", eoh1, emptyHash1);
        assertNotEquals("Null-equality of EndpointKeyHash", eoh1, null);
    }

    @Test
    public void checkOnAttachedCallback() throws IOException {
        KaaClientState state = mock(KaaClientState.class);
        when(state.getEndpointAccessToken()).thenReturn("");

        AttachEndpointToUserCallback listener = mock(AttachEndpointToUserCallback.class);

        DefaultEndpointRegistrationManager manager = new DefaultEndpointRegistrationManager(state, executorContext, null, null);
        manager.setAttachedCallback(null);
        manager.onUpdate(null, null, null, new UserAttachNotification("foo", "bar"), null);
        manager.setAttachedCallback(listener);
        manager.onUpdate(null, null, null, new UserAttachNotification("foo", "bar"), null);

        verify(listener, Mockito.timeout(1000).times(1)).onAttachedToUser("foo", "bar");
        verify(state, Mockito.timeout(1000).times(2)).setAttachedToUser(true);

        manager.setAttachedCallback(null);
        manager.attachUser("externalId", "foo", "bar", null);
        manager.onUpdate(null, null, new UserAttachResponse(SyncResponseResultType.SUCCESS, null, null), null, null);

        manager.setAttachedCallback(listener);
        manager.attachUser("externalId", "foo", "bar", null);
        manager.onUpdate(null, null, new UserAttachResponse(SyncResponseResultType.SUCCESS, null, null), null, null);

        verify(listener, Mockito.timeout(1000).times(1)).onAttachedToUser(anyString(), anyString());
        verify(state, Mockito.timeout(1000).times(4)).setAttachedToUser(true);
    }

    @Test
    public void checkOnDetachedCallback() throws IOException {
        KaaClientState state = mock(KaaClientState.class);
        when(state.getEndpointAccessToken()).thenReturn("");

        DetachEndpointFromUserCallback listener = mock(DetachEndpointFromUserCallback.class);

        DefaultEndpointRegistrationManager manager = new DefaultEndpointRegistrationManager(state, executorContext, null, null);
        manager.setDetachedCallback(null);
        manager.onUpdate(null, null, null, null, new UserDetachNotification("foo"));

        manager.setDetachedCallback(listener);
        manager.onUpdate(null, null, null, null, new UserDetachNotification("foo"));


        verify(listener, Mockito.timeout(1000).times(1)).onDetachedFromUser("foo");
        verify(state, Mockito.timeout(1000).times(2)).setAttachedToUser(false);
    }

}

File: client/client-multi/client-objective-c/KaaTests/event/registration/DefaultEndpointRegistrationManagerTest.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define HC_SHORTHAND
#import <OCHamcrest/OCHamcrest.h>

#define MOCKITO_SHORTHAND
#import <OCMockito/OCMockito.h>

#import <XCTest/XCTest.h>
#import "ExecutorContext.h"
#import "KaaClientState.h"
#import "UserTransport.h"
#import "DefaultEndpointRegistrationManager.h"

#define REQUEST_ID 42

#pragma mark - ConcreteUserAttachDelegate

@interface ConcreteUserAttachDelegate : NSObject <UserAttachDelegate>

@end

@implementation ConcreteUserAttachDelegate

- (void)onAttachResult:(UserAttachResponse *)response {
#pragma unused(response)
}

@end

#pragma mark - DefaultEndpointRegistrationManagerTest

@interface DefaultEndpointRegistrationManagerTest : XCTestCase

@property (nonatomic, strong) id<ExecutorContext> executorContext;
@property (nonatomic, strong) NSOperationQueue *executor;

@end

@implementation DefaultEndpointRegistrationManagerTest

- (void)setUp {
    [super setUp];
    
    self.executorContext = mockProtocol(@protocol(ExecutorContext));
    self.executor = [[NSOperationQueue alloc] init];
    [given([self.executorContext getApiExecutor]) willReturn:self.executor];
    [given([self.executorContext getCallbackExecutor]) willReturn:self.executor];
}

- (void)testCheckUserAttachWithNoDefaultVerifier {
    @try {
        id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
        [given([state endpointAccessToken]) willReturn:@""];
        
        id<UserTransport> transport = mockProtocol(@protocol(UserTransport));
        DefaultEndpointRegistrationManager *manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:transport profileTransport:nil];
        
        [manager attachUserWithId:@"verifierToken" userAccessToken:@"externalId" delegate:nil];
        XCTFail();
    }
    @catch (NSException *exception) {
        NSLog(@"test CheckUserAttachWithNoDefaultVerifier passed");
    }
}

- (void)testCheckUserAttachWithCustomVerifier {
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    [given([state endpointAccessToken]) willReturn:@""];
    
    id<UserTransport> transport = mockProtocol(@protocol(UserTransport));
    DefaultEndpointRegistrationManager *manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:transport profileTransport:nil];
    
    [manager attachUserWithVerifierToken:@"verificationToken" userExternalId:@"externalId" userAccessToken:@"token" delegate:nil];
    [verifyCount(transport, times(1)) sync];
}

- (void)testCheckAttachEndpoint {
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    [given([state endpointAccessToken]) willReturn:@""];
    id<UserTransport> transport = mockProtocol(@protocol(UserTransport));
    DefaultEndpointRegistrationManager *manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:transport profileTransport:nil];
    id<OnAttachEndpointOperationDelegate> delegate = mockProtocol(@protocol(OnAttachEndpointOperationDelegate));
    
    EndpointAccessToken *token1 = [[EndpointAccessToken alloc] initWithToken:@"accessToken1"];
    EndpointAccessToken *token2 = [[EndpointAccessToken alloc] initWithToken:@"accessToken2"];
    
    [manager attachEndpointWithAccessToken:token1 delegate:delegate];
    [manager attachEndpointWithAccessToken:token2 delegate:delegate];
    
    manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:nil profileTransport:nil];
    [manager attachEndpointWithAccessToken:[[EndpointAccessToken alloc] initWithToken:@"accessToken3"] delegate:nil];
    [verifyCount(transport, times(2)) sync];
}

- (void)testcheckDetachEndpoint {
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    [given([state endpointAccessToken]) willReturn:@""];
    id<UserTransport> transport = mockProtocol(@protocol(UserTransport));
    id<OnDetachEndpointOperationDelegate> delegate = mockProtocol(@protocol(OnDetachEndpointOperationDelegate));
    
    DefaultEndpointRegistrationManager *manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:transport profileTransport:nil];
    [manager detachEndpointWithKeyHash:[[EndpointKeyHash alloc] initWithKeyHash:@"keyHash1"] delegate:delegate];
    [manager detachEndpointWithKeyHash:[[EndpointKeyHash alloc] initWithKeyHash:@"keyHash2"] delegate:delegate];
    
    manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:nil profileTransport:nil];
    [manager detachEndpointWithKeyHash:[[EndpointKeyHash alloc] initWithKeyHash:@"keyHash3"] delegate:nil];
    [verifyCount(transport, times(2)) sync];
}

- (void)testCheckAttachUser {
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    [given([state endpointAccessToken]) willReturn:@""];
    
    id<UserTransport> transport = mockProtocol(@protocol(UserTransport));
    DefaultEndpointRegistrationManager *manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:transport profileTransport:nil];
    ConcreteUserAttachDelegate *delegate = [[ConcreteUserAttachDelegate alloc] init];
    [manager attachUserWithVerifierToken:@"externalId" userExternalId:@"userExternalId" userAccessToken:@"userAccessToke" delegate:delegate];
    
    [verifyCount(transport, times(1)) sync];
}

- (void)testCheckWrappers {
    NSString *token1 = @"token1";
    NSString *token2 = @"token2";
    
    EndpointAccessToken *accessToken1 = [[EndpointAccessToken alloc] initWithToken:token1];
    EndpointAccessToken *accessToken1_2 = [[EndpointAccessToken alloc] initWithToken:token1];
    EndpointAccessToken *accessToken2 = [[EndpointAccessToken alloc] initWithToken:token2];
    
    XCTAssertTrue([accessToken1 isEqual:accessToken1]);
    XCTAssertNotEqualObjects(accessToken1, token1);
    XCTAssertEqualObjects([accessToken1 token], [accessToken1 description]);
    XCTAssertEqualObjects(accessToken1, accessToken1_2);
    XCTAssertNotEqualObjects(accessToken1, accessToken2);
    XCTAssertEqual([accessToken1 hash], [accessToken1_2 hash]);
    XCTAssertNotEqual([accessToken1 hash], [accessToken2 hash]);
    
    [accessToken1_2 setToken:token2];
    XCTAssertEqualObjects(accessToken1_2, accessToken2);
    XCTAssertNotEqualObjects(accessToken1, accessToken1_2);
    
    EndpointAccessToken *emptyToken1 = [[EndpointAccessToken alloc] initWithToken:nil];
    EndpointAccessToken *emptyToken2 = [[EndpointAccessToken alloc] initWithToken:nil];
    
    XCTAssertTrue([emptyToken1 isEqual:emptyToken2]);
    XCTAssertEqual([emptyToken1 hash], [emptyToken2 hash]);
    XCTAssertNotEqual(accessToken1, emptyToken2);
    XCTAssertNotNil(accessToken1);
    
    NSString *hash1 = @"hash1";
    NSString *hash2 = @"hash2";
    
    EndpointKeyHash *ekh1 = [[EndpointKeyHash alloc] initWithKeyHash:hash1];
    EndpointKeyHash *ekh1_2 = [[EndpointKeyHash alloc] initWithKeyHash:hash1];
    EndpointKeyHash *ekh2 = [[EndpointKeyHash alloc] initWithKeyHash:hash2];
    
    XCTAssertEqualObjects(ekh1, ekh1);
    XCTAssertNotEqualObjects(ekh1, hash1);
    XCTAssertEqualObjects([ekh1 keyHash], [ekh1 description]);
    XCTAssertEqualObjects(ekh1, ekh1_2);
    XCTAssertNotEqualObjects(ekh1, ekh2);
    XCTAssertNotEqualObjects([ekh1 keyHash], [ekh2 keyHash]);
    XCTAssertEqualObjects([ekh1 keyHash], [ekh1_2 keyHash]);
    
    [ekh1_2 setKeyHash:hash2];
    XCTAssertEqualObjects(ekh1_2, ekh2);
    XCTAssertNotEqualObjects(ekh1_2, ekh1);
    
    EndpointKeyHash *emptyKeyHash1 = [[EndpointKeyHash alloc] initWithKeyHash:nil];
    EndpointKeyHash *emptyKeyHash2 = [[EndpointKeyHash alloc] initWithKeyHash:nil];
    
 //   XCTAssertEqualObjects(emptyKeyHash1, emptyKeyHash2);
    XCTAssertEqual([emptyKeyHash1 hash], [emptyKeyHash2 hash]);
    XCTAssertNotEqualObjects(ekh1, emptyKeyHash1);
    XCTAssertNotNil(ekh1);
}

- (void)testCheckOnAttachedDelegate {
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    [given([state endpointAccessToken]) willReturn:@""];
    UserAttachNotification *attachNotification = [[UserAttachNotification alloc] init];
    attachNotification.userExternalId = @"foo";
    attachNotification.endpointAccessToken = @"bar";
    
    id<AttachEndpointToUserDelegate> delegate = mockProtocol(@protocol(AttachEndpointToUserDelegate));
    
    DefaultEndpointRegistrationManager *manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:nil profileTransport:nil];
    [manager setAttachDelegate:nil];
    [manager onUpdateWithAttachResponses:nil detachResponses:nil userResponse:nil userAttachNotification:attachNotification userDetachNotification:nil];
    [manager setAttachDelegate:delegate];
    [manager onUpdateWithAttachResponses:nil detachResponses:nil userResponse:nil userAttachNotification:attachNotification userDetachNotification:nil];
    
    [NSThread sleepForTimeInterval:1.f];
    [verifyCount(delegate, times(1)) onAttachedToUser:@"foo" token:@"bar"];
    [verifyCount(state, times(2)) setIsAttachedToUser:YES];
    
    [manager setAttachDelegate:nil];
    [manager attachUserWithVerifierToken:@"externalId" userExternalId:@"foo" userAccessToken:@"bar" delegate:nil];
    [manager onUpdateWithAttachResponses:nil detachResponses:nil userResponse:[self getUserAttachResponse] userAttachNotification:nil userDetachNotification:nil];
    
    [manager setAttachDelegate:delegate];
    [manager attachUserWithVerifierToken:@"externalId" userExternalId:@"foo" userAccessToken:@"bar" delegate:nil];
    [manager onUpdateWithAttachResponses:nil detachResponses:nil userResponse:[self getUserAttachResponse] userAttachNotification:nil userDetachNotification:nil];
    
    [NSThread sleepForTimeInterval:1.f];
    [verifyCount(delegate, times(1)) onAttachedToUser:@"foo" token:@"bar"];
    [verifyCount(state, times(4)) setIsAttachedToUser:YES];
}

- (void)testCheckOnDetachedDelegate {
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    [given([state endpointAccessToken]) willReturn:@""];
    UserDetachNotification *detachedNotification = [[UserDetachNotification alloc] init];
    detachedNotification.endpointAccessToken = @"foo";

    id<DetachEndpointFromUserDelegate> delegate = mockProtocol(@protocol(DetachEndpointFromUserDelegate));
    
    DefaultEndpointRegistrationManager *manager = [[DefaultEndpointRegistrationManager alloc] initWithState:state executorContext:self.executorContext userTransport:nil profileTransport:nil];
    [manager setDetachDelegate:nil];
    [manager onUpdateWithAttachResponses:nil detachResponses:nil userResponse:nil userAttachNotification:nil userDetachNotification:detachedNotification];
    
    [manager setDetachDelegate:delegate];
    [manager onUpdateWithAttachResponses:nil detachResponses:nil userResponse:nil userAttachNotification:nil userDetachNotification:detachedNotification];
    
    [NSThread sleepForTimeInterval:1.f];
    [verifyCount(delegate, times(1)) onDetachedEndpointWithAccessToken:@"foo"];
    [verifyCount(state, times(2)) setIsAttachedToUser:NO];
}

- (UserAttachResponse *)getUserAttachResponse {
    UserAttachResponse *response = [[UserAttachResponse alloc] init];
    response.result = SYNC_RESPONSE_RESULT_TYPE_SUCCESS;
    return response;
}

@end
